### DISCUSSION TITLE
usage of the bridge

### DISCUSSION THREAD

**Discussion by TIMESTICKING** [HomeSpan/HomeSpan#476] [General]
I'm new to homekit development, and HomeSpan is really easy to understand to use, but I don't quite understand what's the bridge for while I was trying the samples out, I searched it up says you can add other protocols not only homekit to your devices but how do I manage that with a simple esp32 board

**Comment by TIMESTICKING** [comment]
and also why do you put bridges in those tutorial samples, like LED samples, but actually did nothing with the bridges...

**Comment by HomeSpan** [comment]
Good question -- here's my perspective on the bridge/no-bridge issue:

The idea of a bridge has evolved since Apple first developed HomeKit.  In the original framework, HomeKit devices were either supposed to be connected *directly* to an appliance (such as a standalone smart plug or smart light) or were "hubs" controlling one or more devices that were not necessarily connected (electrically/physically) directly to the device.

For example, if you had a single device that controlled two separate lamps directly connected to the device with two 120V relays, you would simply create a device with two Accessories (no bridge).  On the other hand, if you had a HomeKit device that communicated via bluetooth with two bluetooth-controllable lamps (perhaps in different rooms), you would create a device with three Accessories, the first one being the "bridge."

The protocols above seem to be a convention, rather than a requirement, so in principle it should not matter whether you use a bridge or not.  However, there are some practical differences between the two (some differences arose as Apple released new versions of HomeKit over the past 4 years).

1. Bridges allow for you to add/delete Accessories without confusing the Home App.  This makes using the bridge framework mandatory for a "hub" device in which you can dynamically add/delete light Accessories.  As the changes are made, HomeKit seems to be able to absorb the changes and correctly display the new configuration in the Home App.

2. If you add a second Accessory to a non-bridged configuration, my experience is that HomeKit can often become confused and either not properly render the Accessories in the Home App, or simply refuse to connect (the dreaded "No-Response").  To resolve the issue, you need to unpair the device, type 'H' into the Serial Monitor to force the HomeSpan to generate a new Device ID, and then re-pair the device.

3. When developing HomeSpan applications, it is sometimes common to change the number of Accessories, especially if you are simply experimenting.  Or sometimes you want to switch the Accessory type from a light bulb to a fan, or thermostat, etc.  This occurs, for example, if you are trying out different HomeSpan tutorials.   HomeKit and the Home App seem to be more forgiving of changes (and less likely to get confused) when a device is configured as a bridge.

4. It used to be the case that you could use Characteristic::Name() to name the first Accessory in a device separate from the name of the Device itself.  But Apple changed this a few years ago, and though you can always modify the names of Accessories in the Home App, you can't create a default name for the first Accessory -- it always reverts to the device name, which you specify in `homeSpan.begin()`.  The solution is to use a bridge, where the first Accessory is just the bridge.  The second Accessory is your lamp, fan, etc., and the Home App will honor your choice of Characteristic::Name for Accessories 2, 3, 4, etc.  Just not Accessory 1.

So does this mean you should always use a bridge?  Not necessarily:

* For some reason, it sometimes appears that bridge devices operate slower than non-bridged devices.  I've not done formal testing on this, but it appears to be the case based on anecdotal evidence.  So I tend to *not* use a bridge if I have a device with a fixed number of Accessories that I know will *not* be changing in the future.

*  The Television Service is only displayed in the Home App with a TV Icon if unbridged.  If you put the Television Service in a bridged device, the Home App uses a generic HomeKit icon for the Accessory.

In sum, for the most part it's your choice whether to use a bridge or not, apart from the differences noted above.

**Reply by TIMESTICKING** [reply]
That's great! Really appreciate your explanation, and seems like I did find out the difference that you mentioned in No.3, when I change the number and the type of the accessories, the `name` and the `icon` in the home app changes correspondingly, but I didnt know that's the effort of the bridge. Anyway, this helps me a lot! THX.
